src/traces/scatter/line_points.js

```diff
@@ -387,7 +391,9 @@ module.exports = function linePoints(d, opts) {
 
             clusterRefDist = ptDist(clusterHighPt, clusterStartPt);
 
-            if(clusterRefDist < getTolerance(clusterHighPt, nextPt) * minTolerance) continue;
+            // #3147 - always include the very first and last points for fills
+            if(!(fill && (pti === 0 || len - pti === 1)) &&
```

Comment:
I've never seen `pti === len - 1` written quite that way 🤔

Reply:
yeah, that _was_ weird, fixed in https://github.com/plotly/plotly.js/pull/3696/commits/9e8c3e9783860cc741bb3fb9bd0e9a1aa754ab44